File: CONCEPTS/chaining1.py
import pandas as pd

df = pd.DataFrame([
    {'GAME_ID': 1000, 'TEAM_ID': 100, 'PLAYER_ID': 1, 'FGM': 10, 'FG3M': 12, 'FTM': 12},
    {'GAME_ID': 1000, 'TEAM_ID': 100, 'PLAYER_ID': 2, 'FGM':  4, 'FG3M':  4, 'FTM':  7},
    {'GAME_ID': 1000, 'TEAM_ID': 200, 'PLAYER_ID': 3, 'FGM':  2, 'FG3M':  6, 'FTM':  5},
    {'GAME_ID': 1000, 'TEAM_ID': 200, 'PLAYER_ID': 4, 'FGM':  8, 'FG3M':  2, 'FTM':  7},
    {'GAME_ID': 2000, 'TEAM_ID': 100, 'PLAYER_ID': 1, 'FGM': 10, 'FG3M':  4, 'FTM': 10},
    {'GAME_ID': 2000, 'TEAM_ID': 100, 'PLAYER_ID': 2, 'FGM': 11, 'FG3M':  5, 'FTM':  4},
    {'GAME_ID': 2000, 'TEAM_ID': 300, 'PLAYER_ID': 5, 'FGM':  8, 'FG3M': 10, 'FTM':  9},
    {'GAME_ID': 2000, 'TEAM_ID': 300, 'PLAYER_ID': 6, 'FGM':  7, 'FG3M':  6, 'FTM':  3},
])

def peek(data, message="Data:"):
    print(f"{message} {data}")
    return data

result = (
    df
    .assign(TOTAL_PTS = lambda x: 2*x['FGM'] + 3*x['FG3M'] + x['FTM'])
    .groupby(['GAME_ID', 'TEAM_ID'], as_index=False)
    .agg(TEAM_PTS=('TOTAL_PTS', 'sum'))
    .sort_values(['GAME_ID', 'TEAM_PTS'], ascending=[True, False])
    .pipe(peek, message= "Result")
    
)

File: CONCEPTS/chaining1.R
library(tidyverse)

df <- tribble(
  ~GAME_ID, ~TEAM_ID, ~PLAYER_ID, ~FGM, ~FG3M, ~FTM,
      1000,      100,          1,   10,    12,   12,
      1000,      100,          2,    4,     4,    7,
      1000,      200,          3,    2,     6,    5,
      1000,      200,          4,    8,     2,    7,
      2000,      100,          1,   10,     4,   10,
      2000,      100,          2,   11,     5,    4,
      2000,      300,          5,    8,    10,    9,
      2000,      300,          6,    7,     6,    3
)

peek <- function(df) {
  print(df)
  return(df)
}

df |>
  mutate(TOTAL_PTS = 2*FGM + 3*FG3M + FTM) |>
    group_by(GAME_ID, TEAM_ID) |>
      summarise(TEAM_PTS = sum(TOTAL_PTS), .groups = "drop") |>
        arrange(GAME_ID, desc(TEAM_PTS)) |>
          peek()